reasoner/resolution/framework/Resolver.java

```diff
@@ -313,32 +307,81 @@ public void clear() {
             blocked.clear();
         }
 
-        public Set<Cycle> cycles() {
+        public Set<Cycle> blockingCycles() {
             return iterate(blocked.values()).flatMap(Iterators::iterate).toSet();
         }
 
-        public void block(Request.Factory toBlock, Set<Cycle> blockers) {
-            assert !blockers.isEmpty();
-            assert contains(toBlock);
+        public void block(Request.Factory toBlock, Set<Cycle> cycles) {
+            assert contains(toBlock) && !cycles.isEmpty();
             visits.remove(toBlock);
-            revisits.remove(toBlock);
-            blocked.computeIfAbsent(toBlock, b -> new HashSet<>()).addAll(blockers);
+            Set<Cycle> revisitCycles = revisits.get(toBlock);
+            if (revisitCycles != null) {
+                removeEquivalentCycles(revisitCycles, cycles);
+                if (revisitCycles.isEmpty()) revisits.remove(toBlock);
+            }
+            mergeNewerCycles(blocked.computeIfAbsent(toBlock, b -> new HashSet<>()), cycles);
+        }
+
+        private void removeEquivalentCycles(Set<Cycle> cycles, Set<Cycle> toRemove) {
+            cycles.removeAll(iterate(cycles).filter(cycle -> findEquivalent(toRemove, cycle).isPresent()).toSet());
+        }
+
+        private void retainEquivalentCycles(Set<Cycle> cycles, Set<Cycle> toRetain) {
+            cycles.removeAll(iterate(cycles).filter(cycle -> findEquivalent(toRetain, cycle).isEmpty()).toSet());
+        }
+
+        private void mergeNewerCycles(Set<Cycle> cycles, Set<Cycle> toMerge) {
+            iterate(toMerge).forEachRemaining(m -> {
+                Optional<Cycle> equivalentCycle = findEquivalent(cycles, m);
+                if (equivalentCycle.isPresent() && m.answersSeen() > equivalentCycle.get().answersSeen()) {
+                    cycles.remove(equivalentCycle.get());
+                    cycles.add(m);
+                } else if (equivalentCycle.isEmpty()) {
+                    cycles.add(m);
+                }
+            });
+        }
+
+        private Optional<Cycle> findEquivalent(Set<Cycle> cycles, Cycle cycle) {
+            return iterate(cycles).filter(c -> c.origin().equals(cycle.origin())).first();
         }
 
-        public void unblock(Set<Cycle> cycles) {
+        public Set<Cycle> cyclesToRevisit(Pair<Concludable, ConceptMap> origin, int numAnswers) {
```

Comment:
reading this function alone, without context `cyclesToRevisit` given `origin` doesn't somehow make sense. The fact we're using the output to `revisit` later is not part of the definition of this function, it just does some filtering operation. What do we think a better name is? I think it will make `origin` also become a better name/be renamed too.

Reply:
We're definitely going to replace this, so skipping it